### PR TITLE
Add central state management to aurelia-app

### PR DESCRIPTION
**Pull request by b-yousefi**
Used Store plugin to create a central state management system. Leveraging a single store approach, there is no need to pass unnecessary data and actions between components. Components can access the state by using `connectTo` decorator to connect application's state automatically to a class property. In addition, actions are used to update application state.

### Installed dependencies 

-  aurelia-store
- rxjs 

### Major Changes

- defined `ItemState` to keep the items' state, `items: Item[]`
- defined `getAllItems`, `addItem`, `updateItem`, and `getItemVersions` actions to change `ItemState` 
- add the aurelia-store plugin with initialState of `items: []`; application starts with an empty array of items
- removed actions references and called the defined actions directly where they are needed to manipulate ItemState

### CODE REVIEW

**Review by doktordirk — COMMENTED** [review]
Nice! :1st_place_medal: 

Design-wise some things could be discussed:
- between the options of ease-of-use and concept, i prefer pur functions and an immutable state for store actions as that is easier testable ie to test those one would need to mock fetch. i also wonder about (potentially) jumping back and forth in state history. though i'd need to think about what would be best here. with api calls in the actions, one can't just replay the actions, one the other hand keeping state and backend data synchronized by eg subscribing to certain state changes also would be tricky. did you use [Redux DevTools](https://chrome.google.com/webstore/detail/redux-devtools/lmhkpmbekcpmknklioeibfkpmmfibljd?hl=de-US) on it? such a small project might be worth to make it history aware https://aurelia.io/docs/plugins/store#making-our-app-history-aware. fyi: for easier immutatiblity we use [immerjs](https://github.com/immerjs/immer) atm in other parts. be aware though that aurelia binding to state properties actually changes the state object by adding getters and setters. meaning one can't use immers freezing and can't bind two-way or from-view inputs to state properties but instead needs to either bind to a copy or have change/submit listeners for inputs of state properties
- actions.ts seem superficial to me. yes, one can argue `Api.addItem(item)` is easier to read than `store.dispatch(Mutations.postItem, item);`, but it adds to boillerplate. atm we use like `update = dispatchify(Mutations.getItemVersions)` in add-items-dialog.ts. but then, since imho state should be immutable, it would make sense, to move the fetching to current actions.ts and having current mutations.ts having pur store actions only. 
- you could use dependency injection for Api/Store as we have atm. but then, we don't use any benefit from using dependency injection for Store atm. well, other than injecting it avoids having executed code on script loading in current mutations.ts (the `store.registerAction("getAllItems", getAllItems);` parts, which would then need to move to some constructor functions. something that seems preferable to me

i think playing with an one-step undo/redo could be interesting by eg moving the update item fetch to a state observable should basically be enough to implement that . that approach conflicts with above mentioned option of moving fetching to current actions.ts. fetching and state actions would not be together in that case.

Note: my remarks tends to be sort of condensed. so in case i missed a step which made some parts hard to follow, don't hesitate to tell me

**Comment by mutoprak** [comment]
overall looks pretty good :sparkles: and thanks for working on it :pray: !

my feedback is similar to dirk's but more from a non-framework perspective. I guess we can decouple frontend store and api calls better. Store/State is in a way a browser based repo for us in this case. We should be able to plug in and out the api part. Imagine we want to use this app as a frontend only app which has no connection to a server. As far as I see, here in your changes, the `mutations.ts` class is tightly coupled to API calls. Not sure atm how to do that but I am sure there is a pattern for that in aurelia framework or a best practice.

another discussion point is that we can assume store similar to local git copy. when we make a change, we commit it and once you are happy with your changes, you either squash them and send to the server or push them as separate commits to see the actual history of changes better (sync your local branch with remote branch basically). or when we start editing something, if we think that the branch might have an update (usually we are working on our own branches so there are no changes unsynced if we know no other developer pushed to those branches) we simply pull first and then start commiting. store usage is no different to me honestly.

this pr makes me wonder if we should start focusing on an offline version of our app (as @doktordirk was suggesting back in the days) which simply works as a browser-based no-server app and then we add this app/server connection on top of it (ideally not by changing any existing files but adding new ones and connecting them via configuration to achieve loose coupling).

**Comment by b-yousefi** [comment]
> Nice! 🥇
> 
> Design-wise some things could be discussed:
> 
> * between the options of ease-of-use and concept, i prefer pur functions and an immutable state for store actions as that is easier testable ie to test those one would need to mock fetch. i also wonder about (potentially) jumping back and forth in state history. though i'd need to think about what would be best here. with api calls in the actions, one can't just replay the actions, one the other hand keeping state and backend data synchronized by eg subscribing to certain state changes also would be tricky. did you use [Redux DevTools](https://chrome.google.com/webstore/detail/redux-devtools/lmhkpmbekcpmknklioeibfkpmmfibljd?hl=de-US) on it? such a small project might be worth to make it history aware https://aurelia.io/docs/plugins/store#making-our-app-history-aware. fyi: for easier immutatiblity we use [immerjs](https://github.com/immerjs/immer) atm in other parts. be aware though that aurelia binding to state properties actually changes the state object by adding getters and setters. meaning one can't use immers freezing and can't bind two-way or from-view inputs to state properties but instead needs to either bind to a copy or have change/submit listeners for inputs of state properties
> * actions.ts seem superficial to me. yes, one can argue `Api.addItem(item)` is easier to read than `store.dispatch(Mutations.postItem, item);`, but it adds to boillerplate. atm we use like `update = dispatchify(Mutations.getItemVersions)` in add-items-dialog.ts. but then, since imho state should be immutable, it would make sense, to move the fetching to current actions.ts and having current mutations.ts having pur store actions only.
> * you could use dependency injection for Api/Store as we have atm. but then, we don't use any benefit from using dependency injection for Store atm. well, other than injecting it avoids having executed code on script loading in current mutations.ts (the `store.registerAction("getAllItems", getAllItems);` parts, which would then need to move to some constructor functions. something that seems preferable to me
> 
> i think playing with an one-step undo/redo could be interesting by eg moving the update item fetch to a state observable should basically be enough to implement that . that approach conflicts with above mentioned option of moving fetching to current actions.ts. fetching and state actions would not be together in that case.
> 
> Note: my remarks tends to be sort of condensed. so in case i missed a step which made some parts hard to follow, don't hesitate to tell me

Thank you very much for your insightful comments 🙏

- In fact, I value immutability too. When I was reading the [Store plugin article](https://aurelia.io/docs/plugins/store/), I saw that it had mentioned “Immer.js” library; since I have not used it before and did not know whether you are using it or not, I decided to make use of the spread operator to achieve immutability. I used `concat` method for adding items to the array and as far as I know, it returns a new object. Although there is nothing to enforce it e.g., freeze, I tried not to mutate the state in the application. If I am mistaken please let me know where I had mutated the state. Nevertheless, it is easier to use libraries to take care of object manipulation and keep objects immutable! Therefore, I am going to change the code and make use of “Immer.js” library. I did used Redux DevTools as you guessed and I was planning to make the application history-aware in the next step :smiley: Personally, I think it is better to make changes step by step. Therefore, if you do not mind I would rather postpone it for the next PR.
- I am not sure if I understood the second point well so I am going to check it with you. I think your point is that we should move API calls to actions.ts functions and call the actions.ts functions in mutations.ts to make the functions testable. Then use `dispatchify` for dispatching mutations.ts functions wherever they are needed in the view classes. That seems like a good idea but that way I think the mutations.ts functions would not become pure as they call functions in actions.ts which are not pure as they make HTTP requests🤔. 
- Personally, I like dependency injection too much :blush: but If you believe it does not add many benefits to the project at the moment, I will remove it.   

About the one-step undo/redo, Unfortunately, I did not understand your approach. I think we can achieve the mentioned feature by making the app history-aware. As it is stated in the [Store plugin article](https://aurelia.io/docs/plugins/store#limiting-the-number-of-history-items) it is possible to set a limit for history entries. That way we can increase the limit in the future too.

**Comment by b-yousefi** [comment]
> overall looks pretty good ✨ and thanks for working on it 🙏 !
> 
> my feedback is similar to dirk's but more from a non-framework perspective. I guess we can decouple frontend store and api calls better. Store/State is in a way a browser based repo for us in this case. We should be able to plug in and out the api part. Imagine we want to use this app as a frontend only app which has no connection to a server. As far as I see, here in your changes, the `mutations.ts` class is tightly coupled to API calls. Not sure atm how to do that but I am sure there is a pattern for that in aurelia framework or a best practice.
> 
> another discussion point is that we can assume store similar to local git copy. when we make a change, we commit it and once you are happy with your changes, you either squash them and send to the server or push them as separate commits to see the actual history of changes better (sync your local branch with remote branch basically). or when we start editing something, if we think that the branch might have an update (usually we are working on our own branches so there are no changes unsynced if we know no other developer pushed to those branches) we simply pull first and then start commiting. store usage is no different to me honestly.
> 
> this pr makes me wonder if we should start focusing on an offline version of our app (as @doktordirk was suggesting back in the days) which simply works as a browser-based no-server app and then we add this app/server connection on top of it (ideally not by changing any existing files but adding new ones and connecting them via configuration to achieve loose coupling).

Thank you very much for your comments 🙏 
-	I guess as @doktordirk suggested, by moving API calls to another file i.e., actions.ts, which can by the way be renamed to api to have a more meaningful name, we can have a better design. I will search for Aurelia best practices and see whether I can find better solutions. 
-	When we use Store plugin or any kind of central state management, we can have access to the state wherever we want but the only way to modify the state should be by dispatching actions which result in creation of a new state. Therefore, when we want to update state e.g. item, we don’t work directly on the state but we work with a copy of it and when we hit save the update action is dispatched; only after successful API call the global state changes, a new state is created and observers get notified to re-render. I guess that is similar to the Git analogy that you mentioned, working on a local copy and committing it when you are satisfied with the changes. As for keeping the state up-to-date, now whenever user refreshes the page, items are fetched but they can be changed at any moment and like Git pushes it is up to the user to pull before pushing. Right now, all updates are applied without checking the version to see if the working version is the latest or not. To make it better, we could send the current version number to the server too while updating the item and then there check whether the current version has changed, which means user was not working on the latest version. If so, user could be notified and then decides whether he/she wants to override the changes or not.

**Comment by mutoprak** [comment]
I'm sorry that I reply so late :( Honestly I realized I had only got a notification that you pushed a commit to the PR and had no about the comment you had made. 

Thank you very much for the detailed description! Looks a lot clean this way.

**Comment by b-yousefi** [comment]
It's fine @mutoprak. I thought you might be busy. 

Yeah, I agree, It seems cleaner this way. But I am still not sure if it is the best solution 🤔

**Comment by mutoprak** [comment]
> It's fine @mutoprak. I thought you might be busy.
> 
> Yeah, I agree, It seems cleaner this way. But I am still not sure if it is the best solution thinking

maybe merge this PR and then open another one with the improvements or ideas you have in mind? small steps towards where you want to end up feels better and less complicated than doing them all at once..

**Comment by b-yousefi** [comment]
Okay, I am going to merge this PR and create another one to make the application history-aware 😃
